### ISSUE TITLE
linux concept

### ISSUE THREAD

**Issue by KevinACoder** [KevinACoder/Learning#18]
swap - Swap is a filesystem type – a raw disk partition is formatted as swap
upon system installation. It's treated as second-level RAM by the
OS. When the OS runs out of RAM, it uses swap.

**Comment by KevinACoder** [comment]
Writing
to a flash chip wears it out (it has limited erase-write cycles); hence,
embedded-system designers would rather sacrifice swap and just
use RAM

**Comment by KevinACoder** [comment]
On a VM-based system, we get the following benefits:
- Process-isolation
- The programmer need not worry about physical memory
- Memory-region protection

**Comment by KevinACoder** [comment]
1. What's a core dump anyway?
A core dump is a snapshot of certain dynamic regions (segments) of
the process at the time it crashed (technically, it's a snapshot of
minimally the data and stack segments). The core dump can be
analyzed postmortem using debuggers such as GDB.

**Comment by KevinACoder** [comment]
for GDB to see the symbols (names of functions, variables, line numbers),
one must compile the code with the -g switch (produces debug information).

**Comment by KevinACoder** [comment]
```
/*
Pthread functions may be broken
into two large groupings
-Thread management
    Functions to create, destroy, join, and detach threads. 
-Synchronization
    Functions to manage the synchronization of threads, including mutexes, condition
variables, and barriers.
*/

/*
 Pthread is provided by glibc, but in a separate library and requires explicit linkage
    e.g. gcc -Wall -Werror -pthread beard.c -o beard
    the flag -pthread also setting certain preprocessor defines that control thread safety
*/

/*
thread termination
    - return from start routine "fall off the end of main()"
    - call pthread_exit() 
    - thread cancled by another thread via pthread_cancel()

process termination
    - returns from its main()
    - call exit()
    - executes new binary image via execve()

*/

/*
 RAII(Resource Acquisition Is Initialization) and scoped lock
    - an powerful programming pattern in cpp to deal with resource
        allocation and deallocation by binding the lifetime of resource
        with a scoped object
    - lock is automatically released when mutex fall out of scope, which
        makes unwinding functions and handling errors convenient
*/
class ScopedMutex
{
  public:
    ScopedMutex(pthread_mutex_t &mutex)
        : mutex_(mutex)
    {
        pthread_mutex_lock(&mutex_);
    }
    ~ScopedMutex()
    {
        pthread_mutex_unlock(&mutex_);
    }

  private:
    pthread_mutex_t &mutex_;
};

/*
 Threading model
    - 1:1, translate kernel threads directly to the user-space concept of 
        thread, one-to-one relationship, 'kernel-level threading'
    - N:1, mapping a process with N threads to a single kernel process,
        user-space scheduler to manage threads and a mechanism to catch and
        handle I/O in a nonblocking fashion. 'user-level threading',
        no context switches overhead but also cannot utilize multiple processor and 
        provide trye parallelism
    - N:M, 'hybird threading', user space implement N threads while using M
        native kernel thread, N >= M, mapping the N user threads onto M kernel threads
*/

/*
Coroutines and fibers provide a unit of execution even lighter in 
weight than the thread. The Go programming language provides language-level 
support for coroutine-like constructs in Linux, called Go-routines.
*/

/*
Threading pattern
    -thread per connection, a unit of work is assigned to one thread, thread
        run until completion, uses the kernel to handle the scheduling of work and
        management of I/O.
    when num of in-flight connections reaches limit, connections are queued
        or rejected until connections drop below the limit
    -event-drived pattern, in TPC pattern, most of the threads are spending time waiting,
        decouple the waiting from thread, that is, issue all I\O asynchronously
        and use multiplexed I\O to manage the flow of control in the server    
    request processing -> a series of asynchronous I\O and associated callbacks
        while callbacks can be waited on via multiplexed I\O, when I\O requests
        returned, the event loop hands the callback off to a waiting thread
    event loop, no need to be threaded, but threads added to provide true
        parallelism
*/

/*
race condition
    -a situation in which the unsynchronized access of a shared
        resource by two or more threads leads to erroneous program behavior
    hardware, kernel resource, data in memory(data race)
    -a region of code which should be synchronized, critical region
*/

/*
 Pthread API
*/
#include <pthread.h>
//pthread_attr_t object pointed defines the thread attributes such as
//  stack size, schedulizing parameters, initial detached state
int pthread_create(pthread_t *thread,
                   const pthread_attr_t *attr,
                   void *(*start_routine)(void *),
                   void *arg);
//obtain thread id at runtime
pthread_t pthread_self(void);
//equality operator for thread id
int pthread_equal(pthread_t t1, pthread_t t2);
//terminate a thread deep in a function call stack,
// retval is provided to any thread waiting on the terminating thread’s death
void pthread_exit(void *retval);
//termination of threads by other threads cancellation, return 0 if success, but
//  since actual termination occurs asynchronously, the successful ret flag only
//  indicate cancellation request is processing successfully
int pthread_cancel(pthread_t thread);
//a thread's cancelable state can be enabled or disabled
int pthread_setcancelstate(int state, int *oldstate);
//A thread’s cancellation type is either asynchronous or deferred, asynchronous cancell may kill
//  thread at any point after request made, while deferred only kill thread at specific
//  cancellation points(safe), it's useful when leaving the thread in undefined state (e.g.
//  cancel thread in the middle of a critical region)
int pthread_setcanceltype(int type, int *oldtype);
//allows one thread to block while waiting for the termination of another,
//  joining allows thread to synchronize their execution against the lifetime
//  of other threads,
//return EDEADLK when deadlock was detected
int pthread_join(pthread_t thread, void **retval);
//make the thread no longer joinable
int pthread_detach (pthread_t thread);
/* define and initialize a mutex named `mutex' */
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
int pthread_mutex_init(pthread_mutex_t *restrict mutex,
    const pthread_mutexattr_t *restrict attr);
int pthread_mutex_destroy(pthread_mutex_t *mutex);
//If the mutex is already locked,
//  the calling thread will block until the mutex is unlocked.
int pthread_mutex_lock(pthread_mutex_t *mutex);
int pthread_mutex_unlock (pthread_mutex_t *mutex);
```

**Comment by KevinACoder** [comment]
```
/*
advantages of using thread
1. thread reduce the OS overhead on creating and switching 
    context compare with process
2. threads are tightly coupled therefore easy to share resources
3. thread improve performance by avoid waiting for processing or I/O
4. thread exploits the performance improvement on multiprocessor by
    concurrent
 */

/*
 common issues and concerns of using thread
1. shared resource could lead to defects such as race conditions and
    deadlocks
2. how to design programs based on program understanding
*/

/*
 basic understanding of thread
1. each stack has its own stack
2. the argument (pointer) pass to a thread at creation time is actually
    on thread's stack
3. thread can allocate TLS(thread local storage), which provides small
    pointer arrays to thread. TLS can only be accessed by the specific 
    thread, which assures thread wouldn't modify one another's data
*/

/*
CreateThread
    .start address, 
    .stack size(1 MB by default), one page(4 kB) is committed to
    the stack initially, and more pages committed until reach limit
    .pointer to thread argument
    .return thread's ID and handle
CreateRemoteThread: create a thread in another process

ExitThread
    .a terminated thread will continue to exist until the last handle be 
    closed with CloseHandle()

GetExitCodeThread
    .lpExitCode

SuspendThread ResumeThread
    .every thread has a suspend count, a thread can execute only if the count
    is 0

WaitForMultipleObjects
    .wait for another thread to terminate

TlsAlloc
TlsFree
TlsGetValue(idx)
TlsGetValue(idx, void * val)
    Thread local storage can be used in Dll to replace gloabl storage,
    each calling thread has its own global storage (for thread-safe DLL)

foreground thread (need quick respond) and background thread

SetThreadPriority()
GetThreadPriority()
*/

/*
C library thread safety
c lib was written to operate in single-threaded processes, some functions may
    use static storage to store intermediate results, therefore are not 
    thread-safe when two separate threads simultaneously accessing and modifying
    static storgae
strtok : scan a string to find the next occurrence of a token
    is an example, which maintains persistent state between successive
    function calls

OS solves these problems by supplying thread-safe C lib implementation
    - use specical C functin _beginthreadex() and _endthreadex() instead of 
    CreateThread() and ExitThread() to get rid of the risk of library thread 
    access
    - select Runtime Library, specify Multi-threaded DLL (/MD) 
*/

/*
threading model
    .boss\worker threading model
    .work crew model
    .client\server model, multistage pipeline, work moves from one thread to
        next

thread state
    .running, executing on a processor
    .wait, wait on a nonsignaled handle (thread, process, synchronization obj),
        blocked, sleeping
    .ready, scheduler can put it in running state at any time
    .suspend, a ready thread will not be run
    .terminated

processor affinity
    .limit the processor that can run specific thread, and prevent other
        thread from using this processor


advantages of applying threading model when designing multi-thread program
    .model is well-understand and tested, which avoid many of the mistakes
    .model helps dev obtain the best performance
    .model correspond naturally to structures of programming problem
    .troubleshooting is easy if analyze in terms of models, underlying problem
        (race condition, deadlocks)is seen to violate the basic principles of 
        models
*/

/*
race condition when thread 1 waits for thread 2 before thread 2 is created:
    .create all worker thread in suspended state and resumes only 
        after all worker threads are created
    .the race condition is caused by unsafe assumptions about the progress
        of other thread

thread starvation and priority inversion
    .low-priority thread hold resources required by high-priority thread
    .therefore fairness ensures that all threads will eventually run and
        avoid the two cases


divide and conquer is more than a strategy for algorithm design,
    it can also be key to exploiting multiprocessors
*/

/*
basic principles when design MT prog
    -be certain each distinct child has its own data structures
        pass as thread's function paramter
    -not assume one thread complete before another
    -any thread at any time may preempted or resume execution
    -thread priority can not be a substitute for explicit sync
    -not ignore things that hardly ever happens
    -testing is not enough to prevent code defects, careful design,
        implementation and inspection matters
    -MT prog behavior varies widely among environment, testing on
        variety system
    -keep in mind the size of stack
    -not use thread for activites naturally sequential, not in concurrent
    -keep in mind the overhead of threads(100s), stacks consume virtual memory
        space, and thread context switching could be expensive
    -avoid complexity, keep simple design
*/

/*
Fiber(light-weight threads) - a fiber is a unit of execution that can be scheduled 
    by the application rather than by the kernel.
    -fibers themselves determine which fiber will execute next
    -fibers have independent stacks but run entirely in the context
        of thread which they are scheduled
    -fibers management occurs entirely in user space rather than kernel
    -fibers allow dev to implment co-routine
API
    -ConvertThreadToFiber()  enable fiber operations
    -CreateFiber() create a fiber with a start address, a stack size and 
        a parameter
    -GetFiberData() obtain indivdual fiber data
    -GetCurrentFiber() get fiber identity
    -SwitchToFiber() a running fiber yields control to another
    -DeleteFiber() delete exist fiber along with associated data
*/
```
